### PR TITLE
Improve nullable analysis of local functions

### PR DESCRIPTION
**Pull request by agocke**
This design tries to meld better analysis of nullable reference types in
local functions with performance. To keep the common case one pass,
local functions are analyzed using the starting state that is an
intersection of all the states before its usages (calls, delegate
conversions, etc), but the results of variables made nullable or
non-nullable inside the local function do not propagate to the callers.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -43225,6 +43222,106 @@ static void F4(object? x4)
...
+        {
+            x.ToString();
+            x = null;
+        }
+    }
+}";
+            var comp = CreateCompilation(new[] { source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (11,13): warning CS8602: Dereference of a possibly null reference.
+                //             x.ToString();
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Please put a comment on the appropriate line. I have no idea which `x.ToString()` is causing this.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: And on the other tests.

---
In reply to: [358966254](https://github.com/dotnet/roslyn/pull/40422#discussion_r358966254) [](ancestors = 358966254)

**Comment by 333fred** [comment]
Could you add the test shown in [https://github.com/dotnet/roslyn/issues/40157](https://github.com/dotnet/roslyn/issues/40157)? How does it interact with this change?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -1540,14 +1544,149 @@ public override BoundNode VisitLocal(BoundLocal node)
...
+
+            return null;
+        }
+
+#nullable enable
+
+        private void VisitStatementsWithLocalFunctions(BoundBlock block)
+        {
+            // Since the nullable flow state affects type information, and types can be queried by
+            // the semantic, there needs to be a single flow state input to a local function that
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >semantic [](start = 19, length = 8)

semantic model

**Comment by 333fred** [comment]
Done review pass (commit 1)

**Comment by agocke** [comment]
> Could you add the test shown in #40157? How does it interact with this change?

I didn't try to address that. I can see two possible specs for the behavior in the absence of a call: we could treat captured variables as always being non-null, or we use the annotation.

My preference would be on the first, but I don't think we've made a decision either way.

**Comment by jcouv** [comment]
I'll take a look shortly. 
This PR fixes an issue that is blocking my PR to make `var` nullable (https://github.com/dotnet/roslyn/pull/40755)

**Comment by jcouv** [comment]
@agocke Anything holding us back from merging this? 
This is blocking my PR on making `var` infer nullable types. Thanks

**Comment by jcouv** [comment]
Merged to unblock my other PR. Andy said it's ready to merge

**Comment by agocke** [comment]
SGTM thanks

**Pull request merged**